### PR TITLE
Handle unreachable checks execution scenario

### PR DESCRIPTION
**Pull request by arbulu89**
Handle the unreachable checks execution scenario.

The implementation was not that simple.

- New table which stores the hosts states for each execution added
- The cluster domain changed to store the checks executions in host basis
- Update the projection to store the new hosts state
- Update the frontend

Initially, I thought of adding a new `NotReached` event or similar (actually, some intermediate commits have this code), but eventually I saw that the code was being duplicated, and the only real difference was the name of the event. I thought that it is not enough the hassle.

Please, don't judge the UI, I didn't have enough time to work on it. But I wanted to open the PR on draft so you can have a view while I'm away.

PD: The tests are still missing

Here how it looks like:
unreachable

### CODE REVIEW

**Comment by arbulu89** [comment]
Hey @trento-project/developers , 
Did you have the chance to have a look on this PR?
Should I simply work on the conflicts and tests?

File: lib/trento/application/read_models/host_checks_results_read_model.ex [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  use Ecto.Schema
+
+  import Ecto.Changeset
+
+  @type t :: %__MODULE__{}
+
+  @derive {Jason.Encoder, except: [:__meta__, :__struct__]}
+  @primary_key false
+  schema "hosts_checks_results" do
```
Comment (fabriziosestito) [inline lib/trento/application/read_models/host_checks_results_read_model.ex:14]: hosts checks results or hosts checks execution?

File: lib/trento/domain/cluster/cluster.ex [hunk 2]
```diff
@@ -438,4 +469,42 @@ defmodule Trento.Domain.Cluster do
       %ClusterHealthChanged{cluster_id: cluster_id, health: new_health}
     end
   end
+
+  defp handle_hosts_execution_data(
```
Comment (fabriziosestito) [inline lib/trento/domain/cluster/cluster.ex:473]: nitpick: i would vouch against the usage of the word "data" in the domain logic. Had to scroll to the function to see what it was actually about, maybe we can use a name which is more "domain related", such as "emit_host_execution_completed_event".

File: lib/trento/domain/cluster/cluster.ex [hunk 3]
```diff
@@ -438,4 +469,42 @@ defmodule Trento.Domain.Cluster do
...
+         %{host_id: host_id, reachable: reachable, msg: msg},
+         multi,
+         cluster_id,
+         hosts_executions
+       )
+       when reachable == false do
```
Comment (fabriziosestito) [inline lib/trento/domain/cluster/cluster.ex:479]: no need of using a guard here see above

File: lib/trento/domain/cluster/cluster.ex [hunk 4]
```diff
@@ -438,4 +469,42 @@ defmodule Trento.Domain.Cluster do
       %ClusterHealthChanged{cluster_id: cluster_id, health: new_health}
     end
   end
+
+  defp handle_hosts_execution_data(
+         %{host_id: host_id, reachable: reachable, msg: msg},
+         multi,
```
Comment (fabriziosestito) [inline lib/trento/domain/cluster/cluster.ex:475]: `%{host_id: host_id, reachable: false, msg: msg},`

File: lib/trento/domain/cluster/cluster.ex [hunk 5]
```diff
@@ -438,4 +469,42 @@ defmodule Trento.Domain.Cluster do
...
+    end)
+  end
+
+  defp handle_hosts_execution_data(
+         %{host_id: host_id, reachable: reachable, msg: msg, checks_results: results},
+         multi,
+         cluster_id,
+         _hosts_executions
+       )
+       when reachable == true do
```
Comment (fabriziosestito) [inline lib/trento/domain/cluster/cluster.ex:498]: no need to use a guard here see comments abovev

File: priv/repo/migrations/20220405142819_create_host_checks_results.exs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+defmodule Trento.Repo.Migrations.CreateHostChecksResults do
+  use Ecto.Migration
+
+  def change do
+    create table(:hosts_checks_results, primary_key: false) do
```
Comment (fabriziosestito) [inline priv/repo/migrations/20220405142819_create_host_checks_results.exs:5]: hosts_checks_executions

File: test/trento/application/integration/checks/checks_test.exs [hunk 7]
```diff
@@ -3,6 +3,7 @@ defmodule Trento.Integration.ChecksTest do
   use Trento.DataCase
 
   import Mox
+  import Mock
```
Comment (fabriziosestito) [inline test/trento/application/integration/checks/checks_test.exs:6]: we want to remove mock in a near future and just rely on mox, could you please add an issue about this test to keep track of it?

File: test/trento/application/integration/checks/checks_test.exs [hunk 8]
```diff
@@ -287,4 +300,114 @@ defmodule Trento.Integration.ChecksTest do
 
     assert {:ok, catalog_by_provider} == Checks.get_catalog_grouped_by_provider()
   end
+
+  test "should handle execution started event properly" do
+    with_mock Trento.Commanded, dispatch: fn _, _ -> :ok end do
```
Comment (fabriziosestito) [inline test/trento/application/integration/checks/checks_test.exs:305]: we want to remove mock library soonish and leverage mox for this kind of things, could we add an issue about this test to keep track of the debt?

**Comment by dottorblaster** [comment]
Woh lots of stuff, great! Could you rewrite the history to make the review easier? Thanks!

File: lib/trento/application/read_models/host_checks_results_read_model.ex (hunk 1 above)
Comment (nelsonkopliku) [reply lib/trento/application/read_models/host_checks_results_read_model.ex:14]: Between the two I'd prefer `HostsChecksExecution`.

---

Just as a discussion triggering point:
I cannot also think about the possibility of it being at the moment  just `ChecksExecution` :thinking: 
anyway it is bounded to the context of a host, right?

We can refine also on the language and see whether some extra differentiation is needed.

**Review by nelsonkopliku — COMMENTED** [review]
Great job!

No big thing from my side, @fabriziosestito already pointed out good comments.

Naming is always difficult and good design too :smile:

File: lib/trento/application/read_models/cluster_read_model.ex [hunk 9]
```diff
@@ -24,6 +27,7 @@ defmodule Trento.ClusterReadModel do
     field :details, :map
     field :checks_execution, Ecto.Enum, values: [:not_running, :requested, :running]
 
+    has_many :hosts_executions, HostChecksExecutionsReadModel, foreign_key: :cluster_id
```
[the remark this answers is not in the document]
Comment (arbulu89) [reply lib/trento/application/read_models/cluster_read_model.ex:30]: It was much more hassle to do so, and without big advantages. It forces you to do more hierarchical operations to get the checks (loop through hosts, and them loop through the checks), and as we were already storing the host_id for each check, I thought that it was good enough. In fact, as you commented, it is easier to handle the 2 lists independently in the frontend

File: test/trento/application/integration/checks/checks_test.exs (hunk 8 above)
Comment (arbulu89) [reply test/trento/application/integration/checks/checks_test.exs:306]: At the end added simply a `#TODO` tag. Tech debt issue: https://github.com/trento-project/web/issues/146

File: lib/trento/application/read_models/cluster_read_model.ex (hunk 9 above)
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply lib/trento/application/read_models/cluster_read_model.ex:30]: Perfect, i was just curious about the rationale. at the end read models are for, well, reading, so it make sense to make them closer to the usecase.

**Review by nelsonkopliku — APPROVED** [review]

**Pull request merged**